### PR TITLE
Completes the Context API

### LINKED ISSUE #4: Context API
The client should have a **public** `context()` method that stores a data hash which will be added to notices. There should also be a public `clear()` method to reset the context to the default (`{}`).

- Data should be stored [local to the current thread](https://www.notion.so/Client-Library-Spec-aa891332f7874196aa0695b6d38dca66)
- Data should be sanitized, but not filtered](https://www.notion.so/Client-Library-Spec-aa891332f7874196aa0695b6d38dca66)

```crystal
# Adding context:
Honeybadger.context({ foo: "bar" })

# Clearing context:
Honeybadger.clear()
```

The client should also support passing local context into `Honeybadger.notify` (and similar methods) as an option:

```crystal
Honeybadger.notify(exception, {
  context: {
    user_id: 1
  }
})
```

When global context also exists, the local context should take precedence.

### Examples

- [`Honeybadger.context` in Ruby](https://github.com/honeybadger-io/honeybadger-ruby/blob/2d24a36f5f560c4234f22f59153c80f32e4ce177/lib/honeybadger/agent.rb#L253)
- [`Honeybadger.clear` in Ruby](https://github.com/honeybadger-io/honeybadger-ruby/blob/2d24a36f5f560c4234f22f59153c80f32e4ce177/lib/honeybadger/agent.rb#L259)

### Resources

- [Client Library Spec](https://www.notion.so/honeybadger/Client-Library-Spec-aa891332f7874196aa0695b6d38dca66#d9a1892930694580a9f41eebb99c22bd)
- [Honeybadger Documentation](https://docs.honeybadger.io)
- [Exceptions API Documentation](https://docs.honeybadger.io/api/exceptions.html)
- [Notices JSON Schema](https://github.com/honeybadger-io/honeybadger-schemas/blob/master/exceptions/v1/notices.json)

### PR DESCRIPTION
**Pull request by robacarp**
fixes #4

This is a proof of concept. Since Crystal doesn't currently provide a way to store data in the fibers used by HTTP::Server, the behavior here may be a little different from that of Honeybadger Ruby. Namely:

- When a fiber is spawned, the context of the parent fiber or runtime is not inherited.

As @jgaskins mentioned on #4 Crystal does provide a similar facility to the Honeybadger Context with `Log.context.set` -- this context is similarly not inherited.

I explored the idea of backing the honeybadger context with the Log context, but it seemed like that magic might be difficult to trace if someone did indeed want _not_ to use the contexts together for some reason. Given that the honeybadger data is "more public" than log messages would be, this doesn't feel like a stretch.

Instead I provided a configuration option which enables slurping the Log context into the Honeybadger context just before it's sent to the exception notification API.

Here's a checklist for this to be merge-able:

- [x] decide if the Log.context integration is sufficient
- [x] write some tests around the context api
- [x] refactor / slim down the demo code to be a better example
- new configuration variable:
  - [x] update readme
  - [x] decide if the default value is reasonable
  - [x] ~add an environment variable override?~
- [x] add changelog entry

### CODE REVIEW

File: src/honeybadger.cr [hunk 1]
```diff
@@ -33,6 +34,10 @@ module Honeybadger
     # False = never report data
     property report_data : Bool? = nil
 
+    # Merge the log context into the Honeybadger payload
+    # before sending errors to the reporting API.
+    property merge_log_context : Bool = true
```
Comment (robacarp) [inline src/honeybadger.cr:39]: This is the configuration option which can disable log context data ingestion. I set the default value to be true here because it seems like that's what team Honeybadger would want. This means that any usage of `Log.context.set` already existing in a codebase will automatically populate, so this shard could be more useful straight out of the box -- I'm unclear if the crystal community at large really even knows this feature exists in the Log class.

Comment (joshuap) [reply src/honeybadger.cr:39]: This seems like a reasonable default to me. Log context is a neat feature.

**Review by joshuap — COMMENTED** [review]
Personally, I think this looks like what we want, assuming the `Fiber.current` behavior works for the default path. Reading the example, it appears that you can call `Honeybadger.context({ ... })` in your request handler/routing code and have the context data appear in Honeybadger when an exception is raised; that's a good default. I'm OK with documenting any issues surrounding spawning fibers in the user's code (i.e. if someone is using fibers directly, we can explain that context is not inherited).

@stephendolan do you want to take a glance at this?

File: src/honeybadger.cr (hunk 1 above)
Comment (jgaskins) [reply src/honeybadger.cr:39]: I love this idea. Doesn’t _force_ you to use `Log.context`, but it accommodates it if you choose to. 🙌🏼 I think I’ll incorporate this idea into my Datadog shard. I was going to force using `Log.context`, but I like this approach a lot better!

File: src/honeybadger/context.cr [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    delegate to_s, to: @data
+    forward_missing_to @data
+
+    def self.current : self
+      Fiber.current.honeybadger_context
+    end
+
+    private def set(key, value)
+      self.[key.to_s] = value.to_s
```
Comment (jgaskins) [inline src/honeybadger/context.cr:19]: Was the `.` after `self` a typo? I didn’t even realize you could do this.

```suggestion
      self[key.to_s] = value.to_s
```

File: src/honeybadger/context.cr [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    def initialize
+      @data = Hash(String, String).new
+    end
+
+    delegate to_s, to: @data
+    forward_missing_to @data
+
+    def self.current : self
+      Fiber.current.honeybadger_context
+    end
```
Comment (jgaskins) [inline src/honeybadger/context.cr:16]: This is really slick. One of the things I was most concerned about with the proposed API was that if you did a bunch of work inside child fibers (say, to fan out work to perform it concurrently), you wouldn’t get the parent fiber’s context if an exception were raised in one of the child fibers. For example:

```crystal
def call(http_context)
  channel = Channel(Result?).new

  widgets = WidgetsService.call(http_context.request.query_params[“search”])

  widgets.each do |widget|
    spawn do
      result = process(widget)
    rescue ex
      # FIXME: Missing parent fiber’s context!
      Honeybadger.notify ex
    ensure
      channel.send result
    end
  end

  widgets.each { channel.receive }
end
```

But I think your `Context.current` idea here makes this really simple to handle with 2 additional lines of code:

```diff
 def call(http_context)
   channel = Channel(Result?).new

   widgets = WidgetsService.call(http_context.request.query_params[“search”])

+  parent_context = Context.current
   widgets.each do |widget|
     spawn do
+      Context.current.merge! parent_context
       result = process(widget)
     rescue ex
       Honeybadger.notify ex
     ensure
       channel.send result
     end
   end

   widgets.each { channel.receive }
 end
```

**Review by jgaskins — COMMENTED** [review]
There’s so much to like about this implementation. It seems really easy to use, even if you’re doing some pretty complex things. Thank you for putting so much thought and effort into this!

File: src/honeybadger/context.cr (hunk 2 above)
Comment (robacarp) [reply src/honeybadger/context.cr:19]: It's not a typo but I suppose it is superfluous regardless. Good looking out, that's a tiny detail to notice! Crystal (and ruby) operators are just methods and can be called in strange ways. This is valid in both languages:

```ruby
puts 3.+(1./(4))
```

File: src/honeybadger/context.cr (hunk 3 above)
Comment (robacarp) [reply src/honeybadger/context.cr:16]: @jgaskins that's a good point which I hadn't considered. Perhaps that explicit inheritance is better anyway...or at least that's what we can tell them when they ask!

File: src/honeybadger/context.cr (hunk 2 above)
Comment (jgaskins) [reply src/honeybadger/context.cr:19]: Yeah, I didn’t realize you could just stick the dot in there. 😄 I thought you had to call it similarly to the arithmetic operator methods as you pointed out:

```crystal
self.[]=(key.to_s, value.to_s)
```

**Comment by joshuap** [comment]
Thanks for the review @jgaskins!! 🚀

**Comment by robacarp** [comment]
@joshuap this is ready for a final review. There is some code in here which is difficult to test appropriately, and that resulted in a bug which @stephendolan found in his testing. The resolution is in 74ee5d1 -- I haven't yet been able to reason out straightforward way to write test around this bug+fix. The demo will soft fail without it, but in order to make it hard-fail I'd need to have a way to access the honeybadger api to check that the context didn't make it to the server. I'll think on it tonight and maybe I'll wake up with a solution, but suggestions are welcome.

**Review by joshuap — APPROVED** [review]
Looks good to me, thanks!

Regarding testing the edge case, I'm not sure either; I don't know that it's worth adding some new type of end-to-end integration test just for that case. As long as we have coverage of changes that will crash a user's app, I'd say it's probably not worth the maintenance burden. 🤷
